### PR TITLE
Development

### PR DESCRIPTION
**Pull request by sumaerjolly**

### CODE REVIEW

**Review by AdnanAfsari — CHANGES_REQUESTED** [review]
**Hello @sumaerjolly @em-em-D Great Work :)**

You are almost there but you have to add few more things 

- You have to write test for every methods, you missed few of them and every methods should have 2 3 test case.

- The project should have  a descriptive README.md.

Please complete those  and submit for another review

Happy coding :)

File: README.md [hunk 1]
```diff
@@ -1 +1 @@
-# enumerable
\ No newline at end of file
+In this method we recreated some of the methods in the enumerable model. After creating it we used Rspec to verify whether our methods are correct
```
Comment (mcrd25) [inline README.md:5]: Try adding Github username and links to all contributors
Try adding a title
Try structuring it a bit better using basic Markdown. You can get a cheatsheet [here](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet).

File: spec/enumerable_spec.rb [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    let(:array) { [1,2,3,4] }
+    let(:array1) { ["Ruby", "Javascript", "Python","Javascript"] }
+    let(:array2) { [3,5,7,9] }
+    let(:hash) { {a:"Google", b:"Apple", c:"Facebook"} }
+    let(:hash2) { { "a" => 100, "b" => 200, "c" => 300 } }
+    let(:block) { proc {|x| x} }
+
+    
+    describe "#my_each" do
+        it "Iterates through each element of an array" do
```
Comment (mcrd25) [inline spec/enumerable_spec.rb]: Your description of the test does not match what the test is testing.
This it block actually tests that it returns self, basically all three of your tests whether you change the block passed or not would return self.  So maybe one test like this is fine. What if no block is passed?

If you want to test that it iterates through array etc you may have to do a test like the following:

```ruby
it 'calls the given block once for each element in self' do
      my_each_output = ''
      block = proc { |num| my_each_output += num.to_s }
      array.each(&block)
      each_output = my_each_output.dup
      my_each_output = ''
      array.my_each(&block)
      expect(my_each_output).to eq(each_output)
    end
```
Here are some more ideas:
```ruby
it 'returns original array when no output' do
  expect(test_arr.my_each { test_block }).to eq(test_arr)
end

it 'prints result to stdout from block calculation i * 2' do
  expect { test_arr.my_each { |i| print i * 2 } }.to output(/246/).to_stdout
end
```
These are only ideas of course do not replicate it exactly

File: spec/enumerable_spec.rb [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            expect(array.my_each {|n| n}).to eql(array)
+        end
+        it "Iterates through each key and value of a hash" do
+            expect(hash.my_each {|key,value| key}).to eql(hash)
+        end
+        it "iterates through the array when you pass a block also" do
```
Comment (mcrd25) [inline spec/enumerable_spec.rb]: Technically you've tested this already. The previous tests you did pass a block

File: spec/enumerable_spec.rb [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            expect(array.my_each {block}).to eql(array)
+        end
+    
+    end
+
+    describe "#my_each_with_index" do
+        it "passes the index of each element of an array" do
```
Comment (mcrd25) [inline spec/enumerable_spec.rb]: Same issues as with my_each. 

Try comparing this to original each_With_index_method. You may even do negative tests like if your method does not handle when the block is not given, you expect it to raise an error when the block is not given and not to raise an error when the block is given.


Try to be creative =)

File: spec/enumerable_spec.rb [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            expect(array.my_each_with_index {|index| index}).to eql(0...4) 
+        end
+        it "passes the index of each element of an array" do
+            expect(array1.my_each_with_index {|index| index}).to eql(0...4) 
+        end
+    end 
+
+    describe "#my_select" do
```
Comment (mcrd25) [inline spec/enumerable_spec.rb:40]: Your my_select tests are fine. 

Another thing you can do is test to against official enumera le method like 
```ruby
it 'all above-mentioned tests match the official enumerable method counterpart' do
  expect(test_arr.my_select(&:even?)).to eq(test_arr.select(&:even?))
  expect(test_arr.my_select.is_a?(Enumerator)).to eq(test_arr.select.is_a?(Enumerator))
end
```

File: spec/enumerable_spec.rb [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        it "Returns a new array containing all elements of array on which its called which pass the block test." do
+            expect(array.my_select {|num| num.even?}).to eql ([2,4]) 
+        end
+
+        it "Returns a new hash containing all elements of hash on which its called which pass the block test." do
+            expect(hash2.my_select{|key,value| key > "a" }).to eql ({"b" => 200, "c" => 300})
+        end
+    end
+     
+    describe "#my_all?" do
```
Comment (mcrd25) [inline spec/enumerable_spec.rb:49]: These tests are fine also. But try to consider being more creative and maybe testing against original method

File: spec/enumerable_spec.rb [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        it "returns 'true' on an empty array " do
+            expect([].my_all?).to eql(true)
+        end
+        it "returns 'false' when an 1 element evaluates to 'false' based on the block passed in." do
+            expect(array.my_all? { |n| n > 100}).to eql(false)
+        end
+    end
+
+    describe "#my_any?" do
```
Comment (mcrd25) [inline spec/enumerable_spec.rb:57]: This is fine also

File: spec/enumerable_spec.rb [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        it "returns any element within the array" do
+            expect([].my_any?).to eql(false)
+        end
+        it "returns 'true' any element that passes the block of code" do
+            expect(array.my_any? {|x| x > 0}).to eql(true)
+        end
+    end
+
+    describe "#my_map" do
```
Comment (mcrd25) [inline spec/enumerable_spec.rb:65]: this is fine

File: spec/enumerable_spec.rb [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        it "Returns true if none of the elements in the array don't pass the block of code" do
+            expect(array.my_none? {|n| n > 5}).to eql(true)
+        end
+        it "Returns true if none of the elements in the array don't pass the block of code" do
+            expect(array1.my_none? {|n| n.length > 5}).to eql(false)
+        end
+    end
+
+    describe "#my_count" do
+        it "returns the total number of elements in an array" do
```
Comment (mcrd25) [inline spec/enumerable_spec.rb:90]: Every single one of your test structure is 4 in length. Try mixing it up a little. This does not really tell me your method works. What if your method could only count up to 4? Not that it does but you get what I'm saying?

**Review by mcrd25 — CHANGES_REQUESTED** [review]
# Good job with tests! 😄 

**I'd like for you to address my review comments. In particular, I'd like for you to address:**
- my_each
- my_each_with_index
- my_count
**Please also consider my other comments**

**Another thing I'd like for you to address is that your test variables are hard coded which does not give much variety... try randomizing your tests a bit. Here is a good example... this test changes test input on each run and gives you a better idea that your methods work with different values.**

```ruby
ARRAY_SIZE = 100
LOWEST_VALUE = 0
HIGHEST_VALUE = 9

let(:array) { Array.new(ARRAY_SIZE) { rand(LOWEST_VALUE...HIGHEST_VALUE) } }
  let(:block) { proc { |num| num < (LOWEST_VALUE + HIGHEST_VALUE) / 2 } }

  describe 'my_each' do
    it 'calls the given block once for each element in self' do
      my_each_output = ''
      block = proc { |num| my_each_output += num.to_s }
      array.each(&block)
      each_output = my_each_output.dup
      my_each_output = ''
      array.my_each(&block)
      expect(my_each_output).to eq(each_output)
    end
  end

  describe 'my_each_with_index' do
    it 'calls the given block once for each element in self' do
      my_each_output = ''
      block = proc { |num, idx| my_each_output += "Num: #{num}, idx: #{idx}\n" }
      array.each_with_index(&block)
      each_output = my_each_output.dup
      my_each_output = ''
      array.my_each_with_index(&block)
      expect(my_each_output).to eq(each_output)
    end
  end

  describe 'my_select' do
    it 'returns an array containing all elements of enum for which the given block returns a true value' do
      expect(array.my_select(&block)).to eq(array.select(&block))
    end
  end
```

**You have done a really good job so far! I know your previous reviewer requested only that you add the other tests but I'd really like you to improve your testing in general and have your test do what they are described as doing!  We are almost ready for 🚀**


***If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media.***

**[--Maya](https://github.com/mcrd25)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/)

**Review by JamesOkunlade — APPROVED** [review]
### Good to go guys!!!
 
You have implemented the three tests my_each, my_each_with_index, and my_count as requested by the previous reviewer. 

This project meets the basic requirements however, you should learn to employ the method of best practices the last reviewer as recommended. 

Go ahead and merge to master, close pull request and fill the project completion form for this project.

Good luck with your future projects. :+1::+1:

**Pull request merged**